src/Type/Php/InArrayFunctionTypeSpecifyingExtension.php

```diff
@@ -51,11 +60,21 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
 
 		$specifiedTypes = new SpecifiedTypes();
 
+		if (
+			$arrayType instanceof ConstantArrayType && !$arrayType->isEmpty()
+			&& count(TypeUtils::getConstantScalars($needleType)) === 0 && $arrayValueType->isSuperTypeOf($needleType)->yes()
+		) {
+			// Avoid false-positives with e.g. a string needle and array{'self', string} as haystack
+			// For such cases there seems to be nothing more that we can specify unfortunately
+			return $specifiedTypes;
+		}
+
```

Comment:
this is an edge case I'd really like to not have here. But I didn't find a way to specify anything here, that doesn't get normalized away (e.g. $needle is `'self'|string` which is `string` and the array must have a string value for sure, which it already has). There was initially no test case for this, but it was showing up as error in slevomat-cs.
any other ideas?